File: app/src/main/java/com/onemoresecret/bt/layout/USLayout.java
package com.onemoresecret.bt.layout;

import static com.onemoresecret.bt.KeyboardReport.LEFT_SHIFT;

import com.onemoresecret.bt.KeyboardReport;

public class USLayout extends KeyboardLayout {

    public USLayout() {
        layout.put('a', new KeyboardReport[]{new KeyboardReport(0, KeyboardUsage.KBD_A)});
        layout.put('b', new KeyboardReport[]{new KeyboardReport(0, KeyboardUsage.KBD_B)});
        layout.put('c', new KeyboardReport[]{new KeyboardReport(0, KeyboardUsage.KBD_C)});
        layout.put('d', new KeyboardReport[]{new KeyboardReport(0, KeyboardUsage.KBD_D)});
        layout.put('e', new KeyboardReport[]{new KeyboardReport(0, KeyboardUsage.KBD_E)});
        layout.put('f', new KeyboardReport[]{new KeyboardReport(0, KeyboardUsage.KBD_F)});
        layout.put('g', new KeyboardReport[]{new KeyboardReport(0, KeyboardUsage.KBD_G)});
        layout.put('h', new KeyboardReport[]{new KeyboardReport(0, KeyboardUsage.KBD_H)});
        layout.put('i', new KeyboardReport[]{new KeyboardReport(0, KeyboardUsage.KBD_I)});
        layout.put('j', new KeyboardReport[]{new KeyboardReport(0, KeyboardUsage.KBD_J)});
        layout.put('k', new KeyboardReport[]{new KeyboardReport(0, KeyboardUsage.KBD_K)});
        layout.put('l', new KeyboardReport[]{new KeyboardReport(0, KeyboardUsage.KBD_L)});
        layout.put('m', new KeyboardReport[]{new KeyboardReport(0, KeyboardUsage.KBD_M)});
        layout.put('n', new KeyboardReport[]{new KeyboardReport(0, KeyboardUsage.KBD_N)});
        layout.put('o', new KeyboardReport[]{new KeyboardReport(0, KeyboardUsage.KBD_O)});
        layout.put('p', new KeyboardReport[]{new KeyboardReport(0, KeyboardUsage.KBD_P)});
        layout.put('q', new KeyboardReport[]{new KeyboardReport(0, KeyboardUsage.KBD_Q)});
        layout.put('r', new KeyboardReport[]{new KeyboardReport(0, KeyboardUsage.KBD_R)});
        layout.put('s', new KeyboardReport[]{new KeyboardReport(0, KeyboardUsage.KBD_S)});
        layout.put('t', new KeyboardReport[]{new KeyboardReport(0, KeyboardUsage.KBD_T)});
        layout.put('u', new KeyboardReport[]{new KeyboardReport(0, KeyboardUsage.KBD_U)});
        layout.put('v', new KeyboardReport[]{new KeyboardReport(0, KeyboardUsage.KBD_V)});
        layout.put('w', new KeyboardReport[]{new KeyboardReport(0, KeyboardUsage.KBD_W)});
        layout.put('x', new KeyboardReport[]{new KeyboardReport(0, KeyboardUsage.KBD_X)});
        layout.put('y', new KeyboardReport[]{new KeyboardReport(0, KeyboardUsage.KBD_Y)});
        layout.put('z', new KeyboardReport[]{new KeyboardReport(0, KeyboardUsage.KBD_Z)});

        layout.put('1', new KeyboardReport[]{new KeyboardReport(0, KeyboardUsage.KBD_1)});
        layout.put('2', new KeyboardReport[]{new KeyboardReport(0, KeyboardUsage.KBD_2)});
        layout.put('3', new KeyboardReport[]{new KeyboardReport(0, KeyboardUsage.KBD_3)});
        layout.put('4', new KeyboardReport[]{new KeyboardReport(0, KeyboardUsage.KBD_4)});
        layout.put('5', new KeyboardReport[]{new KeyboardReport(0, KeyboardUsage.KBD_5)});
        layout.put('6', new KeyboardReport[]{new KeyboardReport(0, KeyboardUsage.KBD_6)});
        layout.put('7', new KeyboardReport[]{new KeyboardReport(0, KeyboardUsage.KBD_7)});
        layout.put('8', new KeyboardReport[]{new KeyboardReport(0, KeyboardUsage.KBD_8)});
        layout.put('9', new KeyboardReport[]{new KeyboardReport(0, KeyboardUsage.KBD_9)});
        layout.put('0', new KeyboardReport[]{new KeyboardReport(0, KeyboardUsage.KBD_0)});

        layout.put('!', new KeyboardReport[]{new KeyboardReport(LEFT_SHIFT, KeyboardUsage.KBD_1)});
        layout.put('@', new KeyboardReport[]{new KeyboardReport(LEFT_SHIFT, KeyboardUsage.KBD_2)});
        layout.put('#', new KeyboardReport[]{new KeyboardReport(LEFT_SHIFT, KeyboardUsage.KBD_3)});
        layout.put('$', new KeyboardReport[]{new KeyboardReport(LEFT_SHIFT, KeyboardUsage.KBD_4)});
        layout.put('%', new KeyboardReport[]{new KeyboardReport(LEFT_SHIFT, KeyboardUsage.KBD_5)});
        layout.put('^', new KeyboardReport[]{new KeyboardReport(LEFT_SHIFT, KeyboardUsage.KBD_6)});
        layout.put('&', new KeyboardReport[]{new KeyboardReport(LEFT_SHIFT, KeyboardUsage.KBD_7)});
        layout.put('*', new KeyboardReport[]{new KeyboardReport(LEFT_SHIFT, KeyboardUsage.KBD_8)});
        layout.put('(', new KeyboardReport[]{new KeyboardReport(LEFT_SHIFT, KeyboardUsage.KBD_9)});
        layout.put(')', new KeyboardReport[]{new KeyboardReport(LEFT_SHIFT, KeyboardUsage.KBD_0)});

        layout.put('\\', new KeyboardReport[]{new KeyboardReport(0, KeyboardUsage.KBD_BACKSLASH)});
        layout.put('|', new KeyboardReport[]{new KeyboardReport(LEFT_SHIFT, KeyboardUsage.KBD_BACKSLASH)});

        layout.put('=', new KeyboardReport[]{new KeyboardReport(0, KeyboardUsage.KBD_EQUAL)});
        layout.put('+', new KeyboardReport[]{new KeyboardReport(LEFT_SHIFT, KeyboardUsage.KBD_EQUAL)});

        layout.put(',', new KeyboardReport[]{new KeyboardReport(0, KeyboardUsage.KBD_COMMA)});
        layout.put('<', new KeyboardReport[]{new KeyboardReport(LEFT_SHIFT, KeyboardUsage.KBD_COMMA)});

        layout.put('.', new KeyboardReport[]{new KeyboardReport(0, KeyboardUsage.KBD_DOT)});
        layout.put('>', new KeyboardReport[]{new KeyboardReport(LEFT_SHIFT, KeyboardUsage.KBD_DOT)});

        layout.put('/', new KeyboardReport[]{new KeyboardReport(0, KeyboardUsage.KBD_SLASH)});
        layout.put('?', new KeyboardReport[]{new KeyboardReport(LEFT_SHIFT, KeyboardUsage.KBD_SLASH)});

        layout.put(';', new KeyboardReport[]{new KeyboardReport(0, KeyboardUsage.KBD_SEMICOLON)});
        layout.put(':', new KeyboardReport[]{new KeyboardReport(LEFT_SHIFT, KeyboardUsage.KBD_SEMICOLON)});

        layout.put('\'', new KeyboardReport[]{new KeyboardReport(0, KeyboardUsage.KBD_APOSTROPHE)});
        layout.put('"', new KeyboardReport[]{new KeyboardReport(LEFT_SHIFT, KeyboardUsage.KBD_APOSTROPHE)});

        layout.put('\\', new KeyboardReport[]{new KeyboardReport(0, KeyboardUsage.KBD_BACKSLASH)});
        layout.put('|', new KeyboardReport[]{new KeyboardReport(LEFT_SHIFT, KeyboardUsage.KBD_BACKSLASH)});

        layout.put('[', new KeyboardReport[]{new KeyboardReport(0, KeyboardUsage.KBD_LEFTBRACE)});
        layout.put('{', new KeyboardReport[]{new KeyboardReport(LEFT_SHIFT, KeyboardUsage.KBD_LEFTBRACE)});

        layout.put(']', new KeyboardReport[]{new KeyboardReport(0, KeyboardUsage.KBD_RIGHTBRACE)});
        layout.put('}', new KeyboardReport[]{new KeyboardReport(LEFT_SHIFT, KeyboardUsage.KBD_RIGHTBRACE)});

        layout.put('-', new KeyboardReport[]{new KeyboardReport(0, KeyboardUsage.KBD_MINUS)});
        layout.put('_', new KeyboardReport[]{new KeyboardReport(LEFT_SHIFT, KeyboardUsage.KBD_MINUS)});

        layout.put('\t', new KeyboardReport[]{new KeyboardReport(0, KeyboardUsage.KBD_TAB)});
        layout.put(' ', new KeyboardReport[]{new KeyboardReport(0, KeyboardUsage.KBD_SPACE)});
        layout.put('\r', new KeyboardReport[]{});
        layout.put('\n', new KeyboardReport[]{new KeyboardReport(0, KeyboardUsage.KBD_ENTER)});
    }

    @Override
    public KeyboardReport[] forKey(char c) {
        KeyboardReport[] r = layout.get(c);
        if (r == null) {
            r = layout.get(Character.toLowerCase(c));
            if (r != null && r.length == 1) {
                return new KeyboardReport[]{r[0].addModifier(LEFT_SHIFT)};
            }
        }

        return r;
    }

    @Override
    public String toString() {
        return "English (US)";
    }
}

File: app/src/main/java/com/onemoresecret/bt/layout/USLayout.kt
package com.onemoresecret.bt.layout

import android.util.Log
import com.onemoresecret.bt.KeyModifier
import com.onemoresecret.bt.KeyboardUsage

/**
 * US Keyboard Layout
 */
open class USLayout : KeyboardLayout() {
    init {
        layout['a'] = Stroke()
            .type(KeyboardUsage.KBD_A)
        layout['b'] = Stroke()
            .type(KeyboardUsage.KBD_B)
        layout['c'] = Stroke()
            .type(KeyboardUsage.KBD_C)
        layout['d'] = Stroke()
            .type(KeyboardUsage.KBD_D)
        layout['e'] = Stroke()
            .type(KeyboardUsage.KBD_E)
        layout['f'] = Stroke()
            .type(KeyboardUsage.KBD_F)
        layout['g'] = Stroke()
            .type(KeyboardUsage.KBD_G)
        layout['h'] = Stroke()
            .type(KeyboardUsage.KBD_H)
        layout['i'] = Stroke()
            .type(KeyboardUsage.KBD_I)
        layout['j'] = Stroke()
            .type(KeyboardUsage.KBD_J)
        layout['k'] = Stroke()
            .type(KeyboardUsage.KBD_K)
        layout['l'] = Stroke()
            .type(KeyboardUsage.KBD_L)
        layout['m'] = Stroke()
            .type(KeyboardUsage.KBD_M)
        layout['n'] = Stroke()
            .type(KeyboardUsage.KBD_N)
        layout['o'] = Stroke()
            .type(KeyboardUsage.KBD_O)
        layout['p'] = Stroke()
            .type(KeyboardUsage.KBD_P)
        layout['q'] = Stroke()
            .type(KeyboardUsage.KBD_Q)
        layout['r'] = Stroke()
            .type(KeyboardUsage.KBD_R)
        layout['s'] = Stroke()
            .type(KeyboardUsage.KBD_S)
        layout['t'] = Stroke()
            .type(KeyboardUsage.KBD_T)
        layout['u'] = Stroke()
            .type(KeyboardUsage.KBD_U)
        layout['v'] = Stroke()
            .type(KeyboardUsage.KBD_V)
        layout['w'] = Stroke()
            .type(KeyboardUsage.KBD_W)
        layout['x'] = Stroke()
            .type(KeyboardUsage.KBD_X)
        layout['y'] = Stroke()
            .type(KeyboardUsage.KBD_Y)
        layout['z'] = Stroke()
            .type(KeyboardUsage.KBD_Z)

        layout['1'] = Stroke()
            .type(KeyboardUsage.KBD_1)
        layout['2'] = Stroke()
            .type(KeyboardUsage.KBD_2)
        layout['3'] = Stroke()
            .type(KeyboardUsage.KBD_3)
        layout['4'] = Stroke()
            .type(KeyboardUsage.KBD_4)
        layout['5'] = Stroke()
            .type(KeyboardUsage.KBD_5)
        layout['6'] = Stroke()
            .type(KeyboardUsage.KBD_6)
        layout['7'] = Stroke()
            .type(KeyboardUsage.KBD_7)
        layout['8'] = Stroke()
            .type(KeyboardUsage.KBD_8)
        layout['9'] = Stroke()
            .type(KeyboardUsage.KBD_9)
        layout['0'] = Stroke()
            .type(KeyboardUsage.KBD_0)

        layout['!'] = Stroke()
            .press(KeyModifier.LEFT_SHIFT)
            .type(KeyboardUsage.KBD_1)
        layout['@'] = Stroke()
            .press(KeyModifier.LEFT_SHIFT)
            .type(KeyboardUsage.KBD_2)
        layout['#'] = Stroke()
            .press(KeyModifier.LEFT_SHIFT)
            .type(KeyboardUsage.KBD_3)
        layout['$'] = Stroke()
            .press(KeyModifier.LEFT_SHIFT)
            .type(KeyboardUsage.KBD_4)
        layout['%'] = Stroke()
            .press(KeyModifier.LEFT_SHIFT)
            .type(KeyboardUsage.KBD_5)
        layout['^'] = Stroke()
            .press(KeyModifier.LEFT_SHIFT)
            .type(KeyboardUsage.KBD_6)
        layout['&'] = Stroke()
            .press(KeyModifier.LEFT_SHIFT)
            .type(KeyboardUsage.KBD_7)
        layout['*'] = Stroke()
            .press(KeyModifier.LEFT_SHIFT)
            .type(KeyboardUsage.KBD_8)
        layout['('] = Stroke()
            .press(KeyModifier.LEFT_SHIFT)
            .type(KeyboardUsage.KBD_9)
        layout[')'] = Stroke()
            .press(KeyModifier.LEFT_SHIFT)
            .type(KeyboardUsage.KBD_0)

        layout['\\'] = Stroke()
            .type(KeyboardUsage.KBD_BACKSLASH)
        layout['|'] = Stroke()
            .press(KeyModifier.LEFT_SHIFT)
            .type(KeyboardUsage.KBD_BACKSLASH)

        layout['='] = Stroke()
            .type(KeyboardUsage.KBD_EQUAL)
        layout['+'] = Stroke()
            .press(KeyModifier.LEFT_SHIFT)
            .type(KeyboardUsage.KBD_EQUAL)

        layout[','] = Stroke()
            .type(KeyboardUsage.KBD_COMMA)
        layout['<'] = Stroke()
            .press(KeyModifier.LEFT_SHIFT)
            .type(KeyboardUsage.KBD_COMMA)

        layout['.'] = Stroke()
            .type(KeyboardUsage.KBD_DOT)
        layout['>'] = Stroke()
            .press(KeyModifier.LEFT_SHIFT)
            .type(KeyboardUsage.KBD_DOT)

        layout['/'] = Stroke()
            .type(KeyboardUsage.KBD_SLASH)
        layout['?'] = Stroke()
            .press(KeyModifier.LEFT_SHIFT)
            .type(KeyboardUsage.KBD_SLASH)

        layout[';'] = Stroke()
            .type(KeyboardUsage.KBD_SEMICOLON)
        layout[':'] = Stroke()
            .press(KeyModifier.LEFT_SHIFT)
            .type(KeyboardUsage.KBD_SEMICOLON)

        layout['\''] = Stroke()
            .type(KeyboardUsage.KBD_APOSTROPHE)
        layout['"'] = Stroke()
            .press(KeyModifier.LEFT_SHIFT)
            .type(KeyboardUsage.KBD_APOSTROPHE)

        layout['['] = Stroke()
            .type(KeyboardUsage.KBD_LEFTBRACE)
        layout['{'] = Stroke()
            .press(KeyModifier.LEFT_SHIFT)
            .type(KeyboardUsage.KBD_LEFTBRACE)

        layout[']'] = Stroke()
            .type(KeyboardUsage.KBD_RIGHTBRACE)
        layout['}'] = Stroke()
            .press(KeyModifier.LEFT_SHIFT)
            .type(KeyboardUsage.KBD_RIGHTBRACE)

        layout['-'] = Stroke()
            .type(KeyboardUsage.KBD_MINUS)
        layout['_'] = Stroke()
            .press(KeyModifier.LEFT_SHIFT)
            .type(KeyboardUsage.KBD_MINUS)

        layout['\t'] = Stroke()
            .type(KeyboardUsage.KBD_TAB)
        layout[' '] = Stroke()
            .type(KeyboardUsage.KBD_SPACE)
        layout['\r'] = Stroke() //will be ignored
        layout['\n'] = Stroke()
            .type(KeyboardUsage.KBD_ENTER)
    }

    override fun forKey(c: Char): Stroke? {
        var s = layout[c]

        if (s == null) {
            //lookup lower case character and set it uppercase
            val cLower = c.lowercaseChar()
            s = layout[cLower]

            if (s != null) {
                s = s.toUpper()
                if (s == null) Log.e(
                    TAG,
                    "Cannot capitalize $c, create dedicated Stroke definition"
                )
                return s
            }
        } else {
            return s
        }

        return null
    }

    override fun toString(): String {
        return "English (US)"
    }

    companion object {
        private val TAG: String = USLayout::class.java.getSimpleName()
    }
}
